### PR TITLE
feat: Option to keep only particles with hits in Geant4

### PR DESCRIPTION
**Pull request by benjaminhuth**
Introduces a new option (off by default) to the geant4 simulation that allows to only keep particles that also generated hits.

This is so far not solvable by particles selectors, and also most efficiently implemented directly in the simulation.

I added this mostly because of storage reasons when writing out to file, this reduces the number of particles from O(10^6) to O(10^4) for me.

### CODE REVIEW

File: Examples/Algorithms/Geant4/src/ParticleTrackingAction.cpp [hunk 1]
```diff
@@ -59,8 +59,20 @@ void ActsExamples::ParticleTrackingAction::PostUserTrackingAction(
...
+  const auto barcode = eventData.trackIdMapping.at(aTrack->GetTrackID());
+
+  auto hasHits = eventData.particleHitCount.find(barcode) !=
+                     eventData.particleHitCount.end() and
+                 eventData.particleHitCount.at(barcode) > 0;
+
+  if (not m_cfg.keepParticlesWithoutHits and not hasHits) {
+    auto n = eventData.particlesInitial.erase(
+        ActsExamples::SimParticle{barcode, Acts::PdgParticle::eInvalid});
+    assert(n == 1);
```
Comment (andiwand) [inline Examples/Algorithms/Geant4/src/ParticleTrackingAction.cpp:71]: should we use an error log here?

File: Examples/Python/python/acts/examples/simulation.py [hunk 2]
```diff
@@ -572,10 +572,12 @@ def addGeant4(
     preSelectParticles: Optional[ParticleSelectorConfig] = ParticleSelectorConfig(),
     postSelectParticles: Optional[ParticleSelectorConfig] = None,
     recordHitsOfSecondaries=True,
+    keepParticlesWithoutHits=True,
     outputDirCsv: Optional[Union[Path, str]] = None,
     outputDirRoot: Optional[Union[Path, str]] = None,
     logLevel: Optional[acts.logging.Level] = None,
     killVolume: Optional[acts.Volume] = None,
+    keepParticlesWithoutHits=True,
```
Comment (andiwand) [inline Examples/Python/python/acts/examples/simulation.py:580]: duplicate

File: Examples/Algorithms/Geant4/src/ParticleTrackingAction.cpp (hunk 1 above)
Comment (andiwand) [reply Examples/Algorithms/Geant4/src/ParticleTrackingAction.cpp:71]: the problem I see with this is that in release build without our assert workaround this error case will be silent while the `ACTS_ERROR` will at least show up in the log

Comment (benjaminhuth) [reply Examples/Algorithms/Geant4/src/ParticleTrackingAction.cpp:71]: But thats the thing with all `assert` statements...

Comment (andiwand) [reply Examples/Algorithms/Geant4/src/ParticleTrackingAction.cpp:71]: that is true but I mostly see this in a place where a logger is not available or might be expensive to use

but up to you

**Review by andiwand — APPROVED** [review]

File: Examples/Algorithms/Geant4/src/ParticleTrackingAction.cpp (hunk 1 above)
Comment (benjaminhuth) [reply Examples/Algorithms/Geant4/src/ParticleTrackingAction.cpp:71]: I actually would leave it with the `assert`.

**Pull request merged**